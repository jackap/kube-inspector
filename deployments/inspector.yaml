apiVersion: apps/v1
kind: Deployment
metadata:
  name: inspector-deployment
  labels:
    app: inspector
    version: latest
spec:
  replicas: 1
  selector:
    matchLabels:
      app: inspector
  template:
    metadata:
      labels:
        app: inspector
    spec:
      containers:
      - name: inspector
        image: inspector:1.0.0
        imagePullPolicy: Never
        ports:
        - containerPort: 8081
