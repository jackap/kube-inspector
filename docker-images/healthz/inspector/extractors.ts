
export const getNamespace = (hostname: string|null) => hostname ? hostname.split('.')[2] : "None"
export const getService = (hostname: string|null) => hostname ? hostname.split('.')[1] : "None"
export const getServicesRunningOnNamespaces = (data: Array<NmapResult>) => data.reduce((acc:any,current) => {
        const service = getService(current.hostname)
        const namespace:string = getNamespace(current.hostname)
        if (!namespace || namespace == null) return acc
        return {
            ...acc,
            [namespace]: acc[namespace] ? acc[namespace].concat(service) : [service]
        }
    },{})
