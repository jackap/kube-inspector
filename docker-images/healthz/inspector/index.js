"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
// @ts-ignore
var nmap = require('node-nmap');
var DEFAULT_PORT = 9080;
var ipRange = "10.1.0/24";
// @ts-ignore
nmap.nmapLocation = 'nmap'; //default
var pprint = function (element) { return console.log(JSON.stringify(element, null, '\t')); };
var getNamespace = function (hostname) { return hostname ? hostname.split('.')[2] : "None"; };
var getService = function (hostname) { return hostname ? hostname.split('.')[1] : "None"; };
var getNamespaces = function (data) {
    var namespaces = data.map(function (entry) { return getNamespace(entry.hostname); });
    var distinctNameSpaces = Array.from(new Set(namespaces));
    return distinctNameSpaces;
};
function getServicesRunningOnNamespaces(data) {
    var servicesOnNamespace = data.reduce(function (acc, current) {
        var _a;
        var service = getService(current.hostname);
        var namespace = getNamespace(current.hostname);
        if (!namespace || namespace == null)
            return acc;
        return __assign(__assign({}, acc), (_a = {}, _a[namespace] = acc[namespace] ? acc[namespace].concat(service) : [service], _a));
    }, {});
    return servicesOnNamespace;
}
function processData(data) {
    var podsInDefaultNamespace = data.filter(function (entry) { return entry.hostname != null && entry.hostname.includes("default"); });
    //pprint(getNamespaces(data))
    pprint(getServicesRunningOnNamespaces(data));
}
/*let quickscan = new nmap.QuickScan(ipRange);
quickscan.on('complete', function(data: Array<NmapResult>){
  const hostnames = data.map( (entry:NmapResult) => entry.hostname)
  console.log("Hostnames")
  console.log(hostnames.toString())
  console.log(data)
});

quickscan.on('error', function(error: Error){
  console.log(error);
});

 quickscan.startScan();
*/
var results = [
    { hostname: null,
        ip: '10.1.0.1',
        mac: '76:86:6D:5D:69:C8',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-2.kube-dns.kube-system.svc.cluster.local',
        ip: '10.1.0.2',
        mac: '92:74:21:4F:07:91',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-3.kube-dns.kube-system.svc.cluster.local',
        ip: '10.1.0.3',
        mac: '6E:4F:91:39:E7:48',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-4.istiod.istio-system.svc.cluster.local',
        ip: '10.1.0.4',
        mac: '22:05:5C:2D:0D:D7',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-5.istio-ingressgateway.istio-system.svc.cluster.local',
        ip: '10.1.0.5',
        mac: '52:E0:92:61:02:3F',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-6.istio-egressgateway.istio-system.svc.cluster.local',
        ip: '10.1.0.6',
        mac: '72:FD:78:8A:ED:BF',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-7.zipkin.istio-system.svc.cluster.local',
        ip: '10.1.0.7',
        mac: '16:19:65:59:2D:5C',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-8.grafana.istio-system.svc.cluster.local',
        ip: '10.1.0.8',
        mac: '2E:7E:2C:C6:24:90',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-9.kiali.istio-system.svc.cluster.local',
        ip: '10.1.0.9',
        mac: 'E2:F3:FF:A7:EF:9F',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-10.prometheus.istio-system.svc.cluster.local',
        ip: '10.1.0.10',
        mac: 'DE:E9:F7:B0:D6:7A',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-11.details.default.svc.cluster.local',
        ip: '10.1.0.11',
        mac: '56:73:CF:D6:91:38',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-12.ratings.default.svc.cluster.local',
        ip: '10.1.0.12',
        mac: 'D2:5C:8A:5A:12:53',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-13.reviews.default.svc.cluster.local',
        ip: '10.1.0.13',
        mac: 'FA:14:DF:CB:40:8D',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-14.reviews.default.svc.cluster.local',
        ip: '10.1.0.14',
        mac: 'B2:7F:B6:58:31:6C',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-15.reviews.default.svc.cluster.local',
        ip: '10.1.0.15',
        mac: 'BA:28:62:E3:B3:64',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: '10-1-0-16.productpage.default.svc.cluster.local',
        ip: '10.1.0.16',
        mac: '3A:A7:BC:98:10:0C',
        openPorts: null,
        osNmap: null,
        vendor: undefined },
    { hostname: 'inspector-deployment-695f784cc4-nnlgm',
        ip: '10.1.0.17',
        mac: null,
        openPorts: null,
        osNmap: null,
        vendor: undefined }
];
processData(results);
