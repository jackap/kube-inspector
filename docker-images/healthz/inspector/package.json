{
  "name": "inspector",
  "version": "1.0.0",
  "description": "Test your kubernetes infrastructure",
  "main": "index.ts",
  "scripts": {
    "build": "tsc",
    "start": "tsc && node index.js",
    "test": "mocha -r ts-node/register tests/**/*.test.ts",
    "coverage": "nyc -r lcov -e .ts -x \"*.test.ts\" npm run test"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "dockerode": "^3.2.0",
    "node-nmap": "^4.0.0"
  },
  "devDependencies": {
    "@types/chai": "^4.2.11",
    "@types/mocha": "^7.0.2",
    "chai": "^4.2.0",
    "mocha": "^7.2.0",
    "nyc": "^15.0.1",
    "ts-node": "^8.10.2",
    "tslint": "^6.1.2",
    "typescript": "^3.9.3"
  }
}
