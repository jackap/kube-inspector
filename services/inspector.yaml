apiVersion: v1
kind: Service
metadata:
  name: inspector-service
spec:
  type: NodePort
  selector:
    app: inspector
  ports:
    - port: 8081
      targetPort: 8081
