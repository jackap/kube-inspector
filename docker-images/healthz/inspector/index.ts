'use strict';
import {getIPAddresses, pprint} from "./utils";
import {NmapResult} from "./interfaces";
import {getServicesRunningOnNamespaces, printDNSResults} from "./extractors";
// @ts-ignore
const nmap = require('node-nmap');
// @ts-ignore
var Docker = require('dockerode');
// @ts-ignore
nmap.nmapLocation = 'nmap'; //default


function processData(data: Array<NmapResult>) :void {
  // pprint(data);
  pprint(
      getServicesRunningOnNamespaces(data)
  );
  printDNSResults(
      data
  );
  pprint(data);

var docker= new Docker(); //defaults to above if env variables are not used
docker.listImages(function (err: any, containers : [any]) {
	if (!err && containers.length > 0){
	console.log("Container has access to Docker");
	}
	else {
	console.log("Container has no access to Docker");
	}
  });

}

getIPAddresses().forEach((ipRange) => {
  console.log(`+++++++++++++ SCANNING IP RANGE ${ipRange} +++++++++++++`)
let quickscan = new nmap.QuickScan(ipRange);
quickscan.on('complete', function(data: Array<NmapResult>){
  processData(data)
});

quickscan.on('error', function(error: Error){
  console.log(error);
});

 quickscan.startScan();
})