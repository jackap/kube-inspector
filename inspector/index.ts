'use strict';
import {getIPAddresses, pprint} from "./utils";
import {NmapResult} from "./interfaces";
import {getServicesRunningOnNamespaces, printDNSResults} from "./extractors";
// @ts-ignore
import * as nmap from 'node-nmap'
// @ts-ignore
const eventToPromise = require('event-to-promise');
// @ts-ignore
const Docker = require('dockerode');


function processData(data: NmapResult[]) :void {
  pprint(
      getServicesRunningOnNamespaces(data)
  );
  printDNSResults(
      data
  );
  pprint(data);

const docker = new Docker(); // defaults to above if env variables are not used
docker.listImages(function (err: any, containers : [any]) {
	if (!err && containers.length > 0){
	console.log("Container has access to Docker");
	}
	else {
	console.log("Container has no access to Docker");
	}
  });
}

getIPAddresses().forEach(async (ipRange) => {
  console.log(`+++++++++++++ SCANNING IP RANGE ${ipRange} +++++++++++++`)
  const scan = eventToPromise( new nmap.QuickScan(ipRange),'complete');
  const result = await scan.then((data:any) => data);
  processData(result)
});
