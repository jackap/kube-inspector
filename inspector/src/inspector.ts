"use strict";

import { NmapResult, ProcessedNmapData } from "./interfaces";
import { getIPAddresses, scanIpRange } from "./utils";
import {getNamespaces, getServicesRunningOnNamespaces} from "./extractors/kubernetes";
import {fetchDNSResults} from "./extractors/dns";
import {isDockerEnabled} from "./extractors/docker";

export async function processNmapData(
  data: NmapResult[]
): Promise<ProcessedNmapData> {
  const retval = {
    namespaces: getNamespaces(data).sort(),
    services: getServicesRunningOnNamespaces(data),
    dns: await fetchDNSResults(data),
    options: {
      docker: await isDockerEnabled()
    }
  };

  return retval;
}

export const getInspectorOutput = async () => {
  let retval: any[] = [];
  for (const ipRange of getIPAddresses()) {
    console.log(`SCANNING IP RANGE ${ipRange}`);
    const nMapResult = await scanIpRange(ipRange);
    const scanResult = await processNmapData(nMapResult);
    console.log(scanResult);
    retval = [...retval, { ipRange, ...scanResult }];
  }
  return retval;
};
