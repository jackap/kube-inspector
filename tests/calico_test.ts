import {installInspector, setupTests, waitPodsWithStatus} from "./setup";
import {deleteInspector, inspect} from "./inspector";
import {
    installCalico,
    deleteCalico,
    applyDenyToDefaultNamespace,
    deleteDenyToDefaultNamespace,
    applyDenyToAllNamespaces,
    deleteDenyToAllNamespaces,
} from "./calico";
import {deleteIstio, installIstio} from "./istio";
import {exec} from "child_process";
const K8s = require('k8s');

const kubectl = K8s.kubectl({
    binary: 'kubectl'
    ,version: '/api/v1'
});

const verifyNoActivePods = async (kubectl) => {
    const pods = await kubectl.pod.list();
    if(pods.items.length !== 0){
        console.error(pods.items)
    }
  //  expect(pods.items.length).toBe(0)
};

const withCleanupSingleNamespace = async (caller: () => Promise<void>) => {
    await caller
    await kubectl.command("delete --all pods --namespace=default")
    await kubectl.command("delete --all deployments --namespace=default");
    await kubectl.command("delete --all services --namespace=default")

}

const _10MINUTES = 60 * 20 * 1000
describe('Kubernetes cluster tests with istio and calico', () =>{
    beforeAll(async () => {
        jest.setTimeout(60 * 20 * 1000); // 20 minutes
         await installCalico(kubectl);
    });

    afterEach(async () => {
        // FIXME: verify also that pods are not running in test-namespace
        await verifyNoActivePods(kubectl)
        const testName = expect.getState().currentTestName
        console.log(`${testName} Ended`)
    });
    afterAll(async () => await deleteCalico(kubectl));

    it('Istio bookinfo example on single namespace and calico', async () => {
        try {
            await installIstio(kubectl);
            const inspectorUrl = await installInspector(kubectl);
            await waitPodsWithStatus(kubectl);
            let inspectorResponse;
            inspectorResponse = await inspect(inspectorUrl);
            expect(inspectorResponse.length).toBe(1);
            expect(inspectorResponse[0].services.default).toMatchSnapshot();
            expect(inspectorResponse[0].services['test-namespace']).toBeUndefined();
            expect(inspectorResponse[0].namespaces).toMatchObject(
                ['default',
                    'istio-system',
                    null]);
        }
        finally {
            await deleteIstio(kubectl);
            await deleteInspector(kubectl);
            await waitPodsWithStatus(kubectl, 'Terminating');
            await kubectl.command("delete --all pods --namespace=default")
            await kubectl.command("delete --all deployments --namespace=default");
            await kubectl.command("delete --all services --namespace=default");
        }

    });

    it('Istio bookinfo example on single namespace and calico rules', async () => {

        await installIstio(kubectl);
        const inspectorUrl = await installInspector(kubectl);
        await waitPodsWithStatus(kubectl);
        await applyDenyToDefaultNamespace(kubectl);
        let inspectorResponse;
        try {
            inspectorResponse = await inspect(inspectorUrl);
            expect(inspectorResponse.length).toBe(1);
            expect(inspectorResponse[0].services.default).toBeUndefined();
            expect(inspectorResponse[0].namespaces).toMatchObject(
                ['istio-system',null]);
        } finally {
            await deleteDenyToDefaultNamespace(kubectl)
            await kubectl.command("delete --all services --namespace=default")
            await kubectl.command("delete --all deployments --namespace=default");
            await kubectl.command("delete --all pods --namespace=default")
            await waitPodsWithStatus(kubectl, 'Terminating');
        }

    },_10MINUTES);

    it('Installing Istio bookinfo example on multiple namespaces with calico ' +
        'enabled hides istio and kube-system namespaces', async () => {


            await installIstio(kubectl);
            await installIstio(kubectl, 'test-namespace');
            const inspectorUrl = await installInspector(kubectl);
            await waitPodsWithStatus(kubectl);
            let inspectorResponse;
            inspectorResponse = await inspect(inspectorUrl);
            expect(inspectorResponse.length).toBe(1);
            expect(inspectorResponse[0].services.default).toMatchSnapshot();
            expect(inspectorResponse[0].services['test-namespace']).not.toBeUndefined();
            expect(inspectorResponse[0].services['test-namespace']).toMatchSnapshot();

            expect(inspectorResponse[0].namespaces).toMatchObject(
                ['default',
                    'istio-system',
                    'test-namespace',
                    null]);

        await applyDenyToAllNamespaces(kubectl);
        await waitPodsWithStatus(kubectl);
        inspectorResponse = await inspect(inspectorUrl);
        expect(inspectorResponse.length).toBe(1);
        expect(inspectorResponse[0].services.default).toBeUndefined();
        expect(inspectorResponse[0].services['test-namespace']).toBeUndefined();

        expect(inspectorResponse[0].namespaces).toMatchObject(
            ['istio-system',null]);
        await deleteDenyToAllNamespaces(kubectl);
        await kubectl.command("delete --all services --namespace=default")
        await kubectl.command("delete --all deployments --namespace=default");
        await kubectl.command("delete --all pods --namespace=default")
        await kubectl.command("delete namespace test-namespace").catch()
        await waitPodsWithStatus(kubectl, 'Terminating');
    });


});
