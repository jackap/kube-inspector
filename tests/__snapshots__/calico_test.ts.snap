// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kubernetes cluster tests with istio and calico Installing Istio bookinfo example on multiple namespaces with calico enabled hides istio and kube-system namespaces 1`] = `
Array [
  "details",
  "productpage",
  "ratings",
  "reviews",
  "reviews",
  "reviews",
]
`;

exports[`Kubernetes cluster tests with istio and calico Installing Istio bookinfo example on multiple namespaces with calico enabled hides istio and kube-system namespaces 2`] = `
Array [
  "details",
  "productpage",
  "ratings",
  "reviews",
  "reviews",
  "reviews",
]
`;

exports[`Kubernetes cluster tests with istio and calico Istio bookinfo example on single namespace and calico 1`] = `
Array [
  "details",
  "productpage",
  "ratings",
  "reviews",
  "reviews",
  "reviews",
]
`;
